### ISSUE TITLE
[Tune] Trial Scheduler: Termination Causes GPU to Not Release Memory

### ISSUE THREAD

**Issue by Hamptonjc** [ray-project/ray#23971]
### What happened + What you expected to happen

While running Ray Tune with PyTorch Lightning, CUDA throws a GPU out-of-memory error after several trial scheduler terminations. Removal of the scheduler alleviates the issue. It has been reproduce with both ASHA and Median Stopping Rule.
Using HyperOpt for tuning.

Expectation: Trial scheduler to safely stop trial and release GPU memory. 

### Versions / Dependencies

Docker image: rayproject/ray:3341fa-py38-gpu
ray: 2.0.0.dev0
ray-lightning: 0.3.0
pytorch: 1.11.0
pytorch-lightning: 1.5.10
cudatoolkit: 11.3.1
CUDA Version: 11.5

### Reproduction script

```python

def tune_train(tune_config):

    ray.tune.utils.wait_for_gpu(target_util=0.05)
    
    # Local imports
    import LitModule
    from config import config
    from dataset import LitDataset

    # update config object
    for k,v in tune_config.items():
        try:
            getattr(config, k)
        except:
            raise ConfigurationError(
                f"Hyperparameter {k} does not exist")

        setattr(config, k, v)

    # Instatiate dataset
    dataset = LitDataset(config=config)
    # Instatiate model
    lit_model = LitModule(config=config)
    # Get trial save dir
    trial_dir = ray.tune.get_trial_dir()
    # Callbacks/strategies
    strategy = None

    callbacks = [TuneReportCallback(
                        metrics={"loss": "val_loss"}),
                     pl.callbacks.ModelCheckpoint(
                        monitor='val_loss', dirpath=trial_dir,
                        filename='{epoch:02d}-{val_loss:.10f}',
                        save_top_k=config.SAVE_TOP_K,
                        mode='min', save_last=True)]

    gpu_monitor = pl.callbacks.GPUStatsMonitor()
    callbacks.append(gpu_monitor)

    # Trainer
    trainer = pl.Trainer(max_epochs=config.MAX_EPOCHS,
                         min_epochs=0,
                         callbacks=callbacks,
                         gpus=1,
                         enable_progress_bar = False,
                         logger=loggers,
                         accumulate_grad_batches=config.ACCUMULATE_GRAD_BATCHES,
                         precision=config.TRAINING_PRECISION)

    # train & test
    trainer.fit(model=lit_model, datamodule=dataset)
    trainer.test(model=lit_model, datamodule=dataset)


def run_tuning():
    # Scheduler
    scheduler = ray.tune.schedulers.MedianStoppingRule(
                            max_t=config.MAX_EPOCHS,
                           grace_period=0, hard_stop=True)

    # tune alg
    alg = HyperOptSearch(metric="loss", mode="min")
    
    # specify resources
    resources = {'gpu': 1, 'cpu': 0}

    # Run Tune
    analysis = ray.tune.run(tune_train,
                            search_alg=alg,
                            config=config.TUNE_SEARCH_SPACE,
                            metric="loss",
                            mode="min",
                            resources_per_trial=resources,
                            name=config.RUN_NAME,
                            scheduler=scheduler,
                            num_samples=config.TUNE_N_SAMPLES)



if __name__ == '__main__':
    run_tuning()

```
Eventually, after ~100+ terminations, CUDA throws error. (On GPU with 8G of GRAM)

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by amogkam** [comment]
Hey @Hamptonjc do you mind sharing the code for `LitModule` as well?

**Comment by amogkam** [comment]
In particular, how many workers are you using for data loading? There are known issues with using multiple workers for the Pytorch data loader and GPU memory not being properly cleaned up.

**Comment by Hamptonjc** [comment]
@amogkam The code is closed source. I will try to reproduce with a toy example. Though I am using multiple dataloader workers (4 to be precise), so that might be the issue.

**Comment by amogkam** [comment]
Can you check to see if the same issue happens if you don't set any workers for the dataloader? Thanks!

**Comment by amogkam** [comment]
This comment might also be relevant in explaining what's going: https://github.com/ray-project/ray/issues/23785#issuecomment-1095861996.

You may also want to try setting the `RAY_kill_worker_timeout_milliseconds` environment variable to a larger number (like 10000) prior to running your script.

**Comment by Hamptonjc** [comment]
With no dataloader workers, the issue is gone. Trying RAY_kill_worker_timeout_milliseconds=20000 still resulted in an error (though delayed from where the error typically occurs at around ~120 trial terminations). Currently attempting  RAY_kill_worker_timeout_milliseconds=60000.

Closing, as this seems to be a known issue. @amogkam Thanks!

**Comment by Hamptonjc** [comment]
For sake of completeness, setting RAY_kill_worker_timeout_milliseconds=60000 still resulted in a GPU out-of-memory error.
